The current implementation of wxSocketClient cannot connect asynchronously,

that is, if wxSocketClient::Connect(...,false) is not followed by a call of
wxSocketClient::WaitOnConnect(), the socket will connect but cannot be used
after wxSOCKET_CONNECTION is received (calling WaitOnConnect() later has no
effect).

Fix the bug by remembering wxSOCKET_CONNECTION_FLAG (which is only set once)
until it has been processed and m_connected is set.

Signed-off-by: Anders Larsen <al@alarsen.net>
---

 src/common/socket.cpp |    9 +++++----
 1 files changed, 5 insertions(+), 4 deletions(-)


diff --git a/src/common/socket.cpp b/src/common/socket.cpp
index 3227bdb..8b4a75d 100644
--- a/src/common/socket.cpp
+++ b/src/common/socket.cpp
@@ -1402,13 +1402,12 @@ wxSocketBase::DoWait(long timeout, wxSocketEventFlags flags)
         }
 
         wxSocketEventFlags events;
+        // reset event flags before starting to wait
+        // (keep wxSOCKET_CONNECTION_FLAG since it only gets set once)
+        m_eventsgot &= wxSOCKET_CONNECTION_FLAG;
         if ( eventLoop )
         {
-            // reset them before starting to wait
-            m_eventsgot = 0;
-
             eventLoop->DispatchTimeout(timeLeft);
-
             events = m_eventsgot;
         }
         else // no event loop or waiting in another thread
@@ -1417,6 +1416,7 @@ wxSocketBase::DoWait(long timeout, wxSocketEventFlags flags)
             timeval tv;
             SetTimeValFromMS(tv, timeLeft);
             events = m_impl->Select(flags | wxSOCKET_LOST_FLAG, &tv);
+            events |= m_eventsgot;
         }
 
         // always check for wxSOCKET_LOST_FLAG, even if flags doesn't include
@@ -1438,6 +1438,7 @@ wxSocketBase::DoWait(long timeout, wxSocketEventFlags flags)
         {
             m_connected = true;
             m_establishing = false;
+            m_eventsgot &= ~wxSOCKET_CONNECTION_FLAG;
             rc = true;
             break;
         }
